### PR TITLE
Fatema Alshaikh

### PR DESCRIPTION
**Pull request by FatemaAlshaikh**

### CODE REVIEW

File: CancelBooking UseCase/CancelBooking.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import java.time.LocalDate;
 // Data Layer
 // Represents a reservation with details like customer name, room type, room number, price, CPR, phone number, and booking date.
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/CancelBooking.java:11]:
```suggestion
you can make the getters more secure by marking the fields as (final) and this ensures that once a reservation is created, it can’t be modified. 
Also you have to ensure input values are validated before storing them. The (toStoring) method may reveal too much sensitive information like CPR, you can create more secure version or remove sensitive details.
```

File: CancelBooking UseCase/CancelBooking.java [hunk 2]
```diff
@@ -61,6 +62,7 @@ public String toString() {
     }
 }
 // Manages a collection of reservations, allowing adding, removing, and checking existence of reservations.
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/CancelBooking.java:100]:
```suggestion
the current code is using (ArrayList) which is not thread-safe and consider using (CopyOnWriteArrayList) or synchronised block for critical methods in multi-threaded environment. Also you can use (HashMap) for fast lookups of reservations by room number and CPR combination and this will optimise the performance of operations like exist and remove.
```

File: CancelBooking UseCase/CancelBooking.java [hunk 3]
```diff
@@ -86,6 +88,7 @@ public boolean exists(String roomNumber, String cpr) {
 }
 // Business Layer
 // Provides methods for reservation management, including adding, canceling, and displaying reservations.
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/CancelBooking.java:133]:
```suggestion
the (processRefund) method could be further improved to handle cases like no-show or cancellations after check-in. also instead of having a method that handles both logic and user input/output (like cancelReservation), separate out logic from user-interfacing code.
```

File: CancelBooking UseCase/CancelBooking.java [hunk 4]
```diff
@@ -146,6 +149,7 @@ public void showCancellationPolicy() {
 }
 // Presentation Layer
 // Main class to handle user interaction for making and canceling reservations
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/CancelBooking.java:194]:
```suggestion
instead of dealing with logic and UI together in the "main" method, it is better to separate concerns into distinct methods and keep "main" only for managing control flow. Also you can add input validation for room number and CPR input to ensure they are not empty or malformed.
```

File: CancelBooking UseCase/ReservationTest.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import org.junit.Test;
 import java.time.LocalDate;
 import static org.junit.Assert.*;
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/ReservationTest.java:4]:
```suggestion
the (Reservation) class has getters like (getCustomerName()), but in the test, you are calling (getcustomerName()) and it would be better to ensure consistent naming. Also, you can add test edge cases such as the scenario when invalid data is passed to the constructor.
```

File: CancelBooking UseCase/ReservationRepositoryTest.java [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 import static org.junit.Assert.*;
 
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/ReservationRepositoryTest.java:7]:
```suggestion
you can add edge case for (remove) to test removing a reservation that does not exist. You can also add a test for removing an item when the repository is empty.
```

File: CancelBooking UseCase/ReservationServiceTest.java [hunk 7]
```diff
@@ -6,6 +6,7 @@
 
 import static org.junit.Assert.*;
 
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/ReservationServiceTest.java:9]:
```suggestion
consider adding more scenarios where the user input is invalid or canceled or where the user provides an unexpected response. also it is beneficial to test edge cases like invalid inputs.
```

File: CancelBooking UseCase/CancelBookingTest.java [hunk 8]
```diff
@@ -3,6 +3,7 @@
 import java.io.PrintStream;
 import java.time.LocalDate;
 import static org.junit.Assert.*;
+
```
Comment (Fatema-Mohd) [inline CancelBooking UseCase/CancelBookingTest.java:6]:
```suggestion
you can use a custom PrintStream in the test to capture system output and test more specific content in the console output. Also you can test different scenarios like when there are no reservations.
```

**Review by Fatema-Mohd — CHANGES_REQUESTED** [review]
I hope you develop and improve the code.
Thank you.

**Review by Fatema-Mohd — APPROVED** [review]
Great work!

**Review by Fatema-Mohd — APPROVED** [review]
Good work!
